### ISSUE TITLE
Force a physical axis layout rearrangement

### ISSUE THREAD

**Issue by oliverdutton-iso** [google/jax#23471]
I am trying to optimise a function where I think it'd be much faster if I could force the physical transpose to a pre and op, rather than being fused into the op. 

[optimisation_barrier](https://github.com/google/jax/pull/23466) has been super useful for forcing op ordering, however as JAX works on logical axes the physical axis order ops like transpose can not be forced. It would be useful to me to have a separate op which can force the physical axis ordering of an array?

I know Pallas and custom_call can enforce axis ordering [with operand_layout_constraints], would a no-op custom-call with layout constraint be the easiest way to do it?

For reference, below is a dummy use case, forcing a physical transpose then sum, rather than direct sum

```python
import jax
import jax.numpy as jnp
x = jnp.ones((8192, 4096))

f_transpose = jax.jit(lambda x: (x.swapaxes(0,1)).sum())
f_direct = jax.jit(lambda x: x.sum())

print("Lowered")
print(f_transpose.lower(x).as_text())
print(f_direct.lower(x).as_text())

print("Compiled")
print(f_transpose.lower(x).compile().as_text())
print(f_direct.lower(x).compile().as_text())
```

Output:
```
Lowered
module @jit__lambda_ attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<8192x4096xf32> {mhlo.layout_mode = "default"}) -> (tensor<f32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.transpose %arg0, dims = [1, 0] : (tensor<8192x4096xf32>) -> tensor<4096x8192xf32>
    %1 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %2 = stablehlo.reduce(%0 init: %1) applies stablehlo.add across dimensions = [0, 1] : (tensor<4096x8192xf32>, tensor<f32>) -> tensor<f32>
    return %2 : tensor<f32>
  }
}

module @jit__lambda_ attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<8192x4096xf32> {mhlo.layout_mode = "default"}) -> (tensor<f32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %1 = stablehlo.reduce(%arg0 init: %0) applies stablehlo.add across dimensions = [0, 1] : (tensor<8192x4096xf32>, tensor<f32>) -> tensor<f32>
    return %1 : tensor<f32>
  }
}

Compiled
HloModule jit__lambda_, entry_computation_layout={(f32[8192,4096]{1,0})->f32[]}, allow_spmd_sharding_propagation_to_parameters={true}, allow_spmd_sharding_propagation_to_output={true}

%region_0.4 (Arg_0.5: f32[], Arg_1.6: f32[]) -> f32[] {
  %Arg_0.5 = f32[] parameter(0)
  %Arg_1.6 = f32[] parameter(1)
  ROOT %add.7 = f32[] add(f32[] %Arg_0.5, f32[] %Arg_1.6), metadata={op_name="jit(<lambda>)/jit(main)/reduce_sum[axes=(0, 1)]" source_file="<ipython-input-4-bcc448923e08>" source_line=5}
}

ENTRY %main.9 (Arg_0.1: f32[8192,4096]) -> f32[] {
  %Arg_0.1 = f32[8192,4096]{1,0} parameter(0)
  %constant.2 = f32[] constant(0)
  %reduce-window = f32[256,128]{1,0} reduce-window(f32[8192,4096]{1,0} %Arg_0.1, f32[] %constant.2), window={size=32x32 stride=32x32}, to_apply=%region_0.4
  %reduce-window.1 = f32[8,4]{1,0} reduce-window(f32[256,128]{1,0} %reduce-window, f32[] %constant.2), window={size=32x32 stride=32x32}, to_apply=%region_0.4
  ROOT %reduce.8 = f32[] reduce(f32[8,4]{1,0} %reduce-window.1, f32[] %constant.2), dimensions={1,0}, to_apply=%region_0.4, metadata={op_name="jit(<lambda>)/jit(main)/reduce_sum[axes=(0, 1)]" source_file="<ipython-input-4-bcc448923e08>" source_line=5}
}


HloModule jit__lambda_, entry_computation_layout={(f32[8192,4096]{1,0})->f32[]}, allow_spmd_sharding_propagation_to_parameters={true}, allow_spmd_sharding_propagation_to_output={true}

%region_0.3 (Arg_0.4: f32[], Arg_1.5: f32[]) -> f32[] {
  %Arg_0.4 = f32[] parameter(0)
  %Arg_1.5 = f32[] parameter(1)
  ROOT %add.6 = f32[] add(f32[] %Arg_0.4, f32[] %Arg_1.5), metadata={op_name="jit(<lambda>)/jit(main)/reduce_sum[axes=(0, 1)]" source_file="<ipython-input-4-bcc448923e08>" source_line=6}
}

ENTRY %main.8 (Arg_0.1: f32[8192,4096]) -> f32[] {
  %Arg_0.1 = f32[8192,4096]{1,0} parameter(0)
  %constant.2 = f32[] constant(0)
  %reduce-window = f32[256,128]{1,0} reduce-window(f32[8192,4096]{1,0} %Arg_0.1, f32[] %constant.2), window={size=32x32 stride=32x32}, to_apply=%region_0.3
  %reduce-window.1 = f32[8,4]{1,0} reduce-window(f32[256,128]{1,0} %reduce-window, f32[] %constant.2), window={size=32x32 stride=32x32}, to_apply=%region_0.3
  ROOT %reduce.7 = f32[] reduce(f32[8,4]{1,0} %reduce-window.1, f32[] %constant.2), dimensions={0,1}, to_apply=%region_0.3, metadata={op_name="jit(<lambda>)/jit(main)/reduce_sum[axes=(0, 1)]" source_file="<ipython-input-4-bcc448923e08>" source_line=6}
}
```

**Comment by oliverdutton-iso** [comment]
Have found 'wrap_with_layout_op', will use with the registering primitives tutorial

**Comment by oliverdutton-iso** [comment]
I've written something which runs a transpose which enforces that logical layout on the physical layout

Any advice on whether it will work if the input layout is not major_to_minor in the operand, or is the LayoutConstraint op somehow inferring this correctly?

```python
from jax import core
from jax.interpreters import mlir
from jax._src.lib.mlir import ir
from jax._src.interpreters.mlir import eval_dynamic_shape_as_tensor

transpose_on_device_p = core.Primitive("transpose_on_device")

def transpose_on_device(x, permutation):
  """
  A relative of jax.lax.transpose, but the output physical layout 
  is forced to match the logical permuted layout.
  """
  physical_axis_order = permutation[::-1]
  return transpose_on_device_p.bind(x, physical_axis_order=physical_axis_order).transpose(permutation)

def transpose_on_device_impl(x, physical_axis_order):
  del physical_axis_order
  return x

def transpose_on_device_abstract_eval(xs, physical_axis_order):
  del physical_axis_order
  return core.ShapedArray(xs.shape, xs.dtype)

def transpose_on_device_lowering(ctx, xc, physical_axis_order):
  out_shape = xc.type.shape
  if core.is_constant_shape(out_shape):
    result_shapes = None
  else:
    result_shapes = [eval_dynamic_shape_as_tensor(ctx, out_shape)]

  op = mlir.custom_call('LayoutConstraint', result_types=[xc.type], operands=[xc],
                   api_version=1,
                   result_shapes=result_shapes,
                   # Set operand layouts to anything. XLA will ignore it.
                   operand_layouts=[tuple(range(len(physical_axis_order)))],  # type: ignore
                   # TODO(yashkatariya): Figure out how to pass tiling to the
                   # custom call.
                   result_layouts=[physical_axis_order],
                  )
  return [op.result]
  
transpose_on_device_p.def_impl(transpose_on_device_impl)
transpose_on_device_p.def_abstract_eval(transpose_on_device_abstract_eval)
mlir.register_lowering(transpose_on_device_p, transpose_on_device_lowering, platform='cuda')


import jax
import jax.numpy as jnp
x = jnp.ones((8, 32, 64, 16))

f_transpose = jax.jit(lambda x: transpose_on_device(x, (0,3,2,1)).sum())
f_direct = jax.jit(lambda x: x.sum())

print("Lowered")
print(f_transpose.lower(x).as_text())
print(f_direct.lower(x).as_text())

print("Compiled")
print(f_transpose.lower(x).compile().as_text())
print(f_direct.lower(x).compile().as_text())
```

**Comment by yashk2810** [comment]
We have a layout API in JAX that you can use: See https://github.com/google/jax/blob/ef947a0ce673201b60763fe1f8b26c1c7dcd0fbc/tests/layout_test.py#L4

**Comment by yashk2810** [comment]
So you probably don't need your own primitive to do this?
